### ISSUE TITLE
Trait method not resolved when implemented on &mut Vec

### ISSUE THREAD

**Issue by panstromek** [intellij-rust/intellij-rust#8553]
## Environment

* **IntelliJ Rust plugin version:** 0.4.164.4409-213
* **Rust toolchain version:** 1.60.0-nightly (27f5d830e 2022-02-02) x86_64-unknown-linux-gnu
* **IDE name and version:** CLion 2021.3.3 (CL-213.6777.58)
* **Operating system:** Windows 10 10.0
* **Macro expansion engine:** new
* **Name resolution engine:** new
* **Additional experimental features:** org.rust.cargo.features.settings.gutter, org.rust.cargo.evaluate.build.scripts, org.rust.macros.proc

## Problem description

Seems like implementing the trait for the reference is the culprit. If I implement the trait for owned value, it's resolved correctly.

Note that this is very old problem, I just didn't report it. It doesn't seem like it's related to new resolution engine (it reproduces on the old one, too).

## Steps to reproduce
Try this code.

```rust

pub trait SomeTrait<T> {
    fn get_some(self) -> T;
}

impl<'a, T> SomeTrait<&'a mut T> for &'a mut Vec<T> {
    fn get_some(self) -> &'a mut T {
        todo!()
    }
}

fn hey() {
    let x = vec![3].get_some(); // .get_some() method is not resolved here, go to definition doesn't work
}

```

**Comment by panstromek** [comment]
There seems to be some progress here, the case that I initially find it on is fixed now, but the one in the reproducer still doesn't work.
